torch/nn/modules/_functions.py

```diff
@@ -15,23 +16,18 @@ def forward(self, input, weight, bias, running_mean, running_var, eps, momentum,
         # calculate mean/invstd for input.
         mean, invstd = torch.batch_norm_stats(input, eps)
 
-        count_all = torch.empty(world_size, 1, dtype=count.dtype, device=count.device)
-        mean_all = torch.empty(world_size, mean.size(0), dtype=mean.dtype, device=mean.device)
-        invstd_all = torch.empty(world_size, invstd.size(0), dtype=invstd.dtype, device=invstd.device)
-
-        count_l = list(count_all.unbind(0))
-        mean_l = list(mean_all.unbind(0))
-        invstd_l = list(invstd_all.unbind(0))
-
-        # using all_gather instead of all reduce so we can calculate count/mean/var in one go
-        count_all_reduce = torch.distributed.all_gather(count_l, count, process_group, async_op=True)
-        mean_all_reduce = torch.distributed.all_gather(mean_l, mean, process_group, async_op=True)
-        invstd_all_reduce = torch.distributed.all_gather(invstd_l, invstd, process_group, async_op=True)
-
-        # wait on the async communication to finish
-        count_all_reduce.wait()
-        mean_all_reduce.wait()
-        invstd_all_reduce.wait()
+        num_channels = input.shape[1]
+        # C, C, 1 -> (2C + 1)
+        combined = torch.cat([mean, invstd, count], dim=0)
+        # world_size * (2C + 1)
+        combined_list = [
+            torch.empty_like(combined) for k in range(world_size)
+        ]
+        # Use allgather instead of allreduce since I don't trust in-place operations ..
```

Comment:
@vkuzo Wondering if there is a better reason than this not to use allreduce :) We use in-place allreduce operations a lot and in general its much faster than allgather.

Reply:
You'd need to do a weighted sum here and not just all-reduce. Here the amount of data transfered is very small, so you'd be limited just by the latency of operation which should be approximately the same for allreduce and allgather.

Reply:
I'm definitely not an expert in this - after finding that this was a bottleneck, for this diff I took the approach of minimizing risk and copying the battle tested piece of the corresponding `detectron2` layer (including this comment :) ).

do we have an estimate of performance savings of using allreduce? We can update it if it's worth it.

Reply:
@ngimel Was looking through this code again and wasn't sure about the weighted all-reduce part. Don't we need to find the global mean and variance for SyncBN and that can be done using all-reduce?